server/grpc_service.go

```diff
@@ -1969,28 +1969,44 @@ func (s *GrpcServer) WatchGlobalConfig(req *pdpb.WatchGlobalConfigRequest, serve
 	// If the revision is compacted, will meet required revision has been compacted error.
 	// - If required revision < CompactRevision, we need to reload all configs to avoid losing data.
 	// - If required revision >= CompactRevision, just keep watching.
-	watchChan := s.client.Watch(ctx, configPath, clientv3.WithPrefix(), clientv3.WithRev(revision))
+	// Use WithPrevKV() to get the previous key-value pair when get Delete Event.
+	watchChan := s.client.Watch(ctx, configPath, clientv3.WithPrefix(), clientv3.WithRev(revision), clientv3.WithPrevKV())
 	for {
 		select {
 		case <-ctx.Done():
 			return nil
 		case res := <-watchChan:
-			if revision < res.CompactRevision {
-				if err := server.Send(&pdpb.WatchGlobalConfigResponse{
-					Revision: res.CompactRevision,
-					Header: s.wrapErrorToHeader(pdpb.ErrorType_DATA_COMPACTED,
-						fmt.Sprintf("required watch revision: %d is smaller than current compact/min revision. %d", revision, res.CompactRevision)),
-				}); err != nil {
+			if res.Err() != nil {
+				var resp pdpb.WatchGlobalConfigResponse
+				if revision < res.CompactRevision {
+					resp.Header = s.wrapErrorToHeader(pdpb.ErrorType_DATA_COMPACTED,
+						fmt.Sprintf("required watch revision: %d is smaller than current compact/min revision %d.", revision, res.CompactRevision))
+				} else {
+					resp.Header = s.wrapErrorToHeader(pdpb.ErrorType_UNKNOWN,
+						fmt.Sprintf("watch channel meet other error %s.", res.Err().Error()))
+				}
+				if err := server.Send(&resp); err != nil {
 					return err
 				}
+				return res.Err()
 			}
 			revision = res.Header.GetRevision()
 
 			cfgs := make([]*pdpb.GlobalConfigItem, 0, len(res.Events))
 			for _, e := range res.Events {
 				// Since item value needs to support marshal of different struct types,
 				// it should be set to `Payload bytes` instead of `Value string`.
-				cfgs = append(cfgs, &pdpb.GlobalConfigItem{Name: string(e.Kv.Key), Payload: e.Kv.Value, Kind: pdpb.EventType(e.Type)})
+				switch e.Type {
+				case clientv3.EventTypePut:
+					cfgs = append(cfgs, &pdpb.GlobalConfigItem{Name: string(e.Kv.Key), Payload: e.Kv.Value, Kind: pdpb.EventType(e.Type)})
+				case clientv3.EventTypeDelete:
+					if e.PrevKv != nil {
+						cfgs = append(cfgs, &pdpb.GlobalConfigItem{Name: string(e.Kv.Key), Payload: e.PrevKv.Value, Kind: pdpb.EventType(e.Type)})
+					} else {
```

Comment:
I think in this case we should just skip this message because prev-kv is compacted means their must be a delete event before this one, that means this is just a duplicated event, so we can just ignore it.